### ISSUE TITLE
Feature request: Preserve MacOS symlink perms on Linux via xattr...

### ISSUE THREAD

**Issue by patniemeyer** [WayneD/rsync#443]
As you probably know, MacOS handles symlinks in a nonstandard way, allowing them to have distinct permissions from their target file via `lstat()` and `lchmod()`, whereas linux systems always report 777 for link perms.  This can result in rsync perpetually seeing permission differences on symlinks when going between MacOS and Linux.

I'm wondering if there would be interest in supporting an option to store these (extra) symlink permissions in extended attributes on linux, allowing them to be preserved in a round trip.

Although I am not familiar with the rsync codebase I would be willing to take a shot at submitting a PR for this.  But if this is not something you think is in scope for rsync please let me know.  Thanks!

**Comment by patniemeyer** [comment]
I came across this old discussion in which you seem to endorse a simpler solution of just ignoring the permissions on symlinks when appropriate.  Was there ever any work on that?

[](https://rsync.samba.narkive.com/700SrOfn/symlinks-in-os-x-10-4-1#post4)

**Comment by devZer0** [comment]
wow, 19 year old issue , told to be fixed  (but still unfixed) :D

i backup macos on regular interval and each rsync run will retransfer/recreate more then 100000 symlinks  on each run, so in wonder how we can avoid this unnecessary action

the problem with this is that rsync log becomes pretty useless, as you have nothing but changes in that log because of permission diff - and won't see what's really transferred

**Comment by devZer0** [comment]
while thinking about ading feature like --no-link-perms disable link check on symlink permissions-  we also could do a runtime check for existance of lchmod,  and if it's missing, skip comparison of symlink permissions. so we could handle this issue automatically

**Comment by patniemeyer** [comment]
I agree  - using rsync on MacOS can be very messy due to this issue.  I'm still willing to help with this - Doing the simple switch to disable the link stat checking seems trivial or we could attempt to round trip the perms through extended attributes.

**Comment by devZer0** [comment]
i have made this patch and it works for me. 

would someone like to review / test ?

also sent this to mailing list

https://github.com/devZer0/rsync-workingcopy/commit/7a8d295e531627e0a3d9e41ceedf10ab9be7fb5e

**Comment by tridge** [comment]
some discussion ideas on how to test from a thread on the rsync discord:

or another idea, rsync has a --fake-super option that stashes permissions in xattrs. So we could pre-create some xattrs with interesting symlink permissions, then do a linux <-> linux transfer where one side is providing symlink permissions via xattrs and the other side using the normal filesystem

that may actually be a viable option for linux based backup servers for macos machines, store the extra permissions using the fake-super xattrs. That way we don't need to just discard, the linux backup server could actually properly store the permissions ready for a restore.

**Comment by devZer0** [comment]
i have added --fake-super to my backup script and for my suprise, now symlinks like

etc -> private/etc

are being turned into ordinary file "etc" with the contents "private/etc"

turns out this is normal (but unexpected) behaviour with fake-super (https://lists.samba.org/archive/rsync/2018-February/031501.html?utm_source=chatgpt.com ) ,  but imho way over the top for my problem - as i'm trying to fix the issue of storing permissions for symlinks and not turning symlinks into files , where the filesystem does support symlinks. 

i think we need a more granular approach here to fix this.

what about the idea of just trying to use --fake-super feature when any syscall for creating link or storing permission goes wrong, and only then fallback to use the alternative method of storing that information?
